janus/lib/client/jsx/janus-admin.jsx

```diff
@@ -45,58 +109,118 @@ const Projects = ({projects}) => (
 const postAddProject = (project) => json_post('/add_project', project);
 
 const NewProject = ({retrieveAllProjects}) => {
-  let [ newproject, setNewProject ] = useState({})
-  let [ error, setError ] = useState(null);
-
-  return <div id='new-project'>
-    <div className='title'>New Project</div>
-    { error && <div className='error'>Error: {error}</div> }
-    <div className='item'>
-      <div className='cell'>
-        <TextField placeholder='project_name'
-          value={ newproject.project_name || '' }
-          onChange={ (e) => setNewProject({ ...newproject, project_name: e.target.value }) }/>
-      </div>
-      <div className='cell'>
-        <TextField
-          placeholder='Project Title'
-          value={ newproject.project_name_full || '' }
-          onChange={ (e) => setNewProject({ ...newproject, project_name_full: e.target.value }) }/>
-      </div>
-      <div className='cell submit'>
-        <Icon className='approve' icon='magic' onClick={
-          () => postAddProject( newproject ).then(
-            ()=> {
-              retrieveAllProjects();
-              setNewProject({});
-              setError(null);
+  let [newproject, setNewProject] = useState({});
+  let [error, setError] = useState(null);
+
+  return (
+    <div id='new-project'>
+      <div className='title'>New Project</div>
+      {error && <div className='error'>Error: {error}</div>}
+      <div className='item'>
+        <div className='cell'>
+          <TextField
+            placeholder='project_name'
+            value={newproject.project_name || ''}
+            onChange={(e) =>
+              setNewProject({...newproject, project_name: e.target.value})
+            }
+          />
+        </div>
+        <div className='cell'>
+          <TextField
+            placeholder='Project Title'
+            value={newproject.project_name_full || ''}
+            onChange={(e) =>
+              setNewProject({...newproject, project_name_full: e.target.value})
             }
-          ).catch(
-            e => e.then( ({error}) => setError(error) )
-          )
-        }/>
+          />
+        </div>
+        <div className='cell submit'>
+          <Icon
+            className='approve'
+            icon='magic'
+            onClick={() =>
+              postAddProject(newproject)
```

Comment:
Hmm...
So firstly, it seems you can easily multi submit here right?  And in that case, which error or not error you see becomes a race condition dependent on network conditions.

This gets even weirder because of the setNewProject({}) which can actually clear project details after the button is clicked at a non deterministic time.

Again, I will always nit the issue of double or multiple submit in forms like this :P.  One alternative is to add additional state, "loading" which only allows one request to be submitted at a time.  There is incidentally already a tool for this in `useAsyncWork` which handles both cancellability and loading state.

Reply:
Ah, thanks, I'll try to pay more attention to this in the future...